### PR TITLE
[SPARK-9034] [SQL] Reflect field names defined in GenericUDTF

### PR DESCRIPTION
**Pull request by navis**
Hive GenericUDTF#initialize() defines field names in a returned schema though,
the current HiveGenericUDTF drops these names.
We might need to reflect these in a logical plan tree.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -148,7 +148,7 @@ class Analyzer(
           child match {
             case _: UnresolvedAttribute => u
             case ne: NamedExpression => ne
-            case g: Generator if g.resolved && g.elementTypes.size > 1 => MultiAlias(g, Nil)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I am not sure if we have to remove the guard `g.elementTypes.size > 1`, can you describe it a little bit?

**Comment by chenghao-intel** [comment]
I'd like to see this will be fixed as it's a long time bug I know. Can you bring it back? It general looks good to me, but I leave a question also in the code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (navis) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Now generator provides default alias for output column, we should ask it in any case.

**Comment by navis** [comment]
Rebased to trunk

**Comment by chenghao-intel** [comment]
LGTM, /cc @liancheng @cloud-fan for taking another look.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -1119,7 +1119,8 @@ class DataFrame private[sql](
   def explode[A <: Product : TypeTag](input: Column*)(f: Row => TraversableOnce[A]): DataFrame = {
     val schema = ScalaReflection.schemaFor[A].dataType.asInstanceOf[StructType]
 
-    val elementTypes = schema.toAttributes.map { attr => (attr.dataType, attr.nullable) }
+    val elementTypes = schema.toAttributes.map {
+      attr => (attr.dataType, attr.nullable, Some(attr.name)) }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: As we specify the names here, we don't need to pass names to `Generate` as `UnresolvedAttribute` at the end of this method anymore.

Comment (navis) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Can I understand it's enough to return this?
```
Generate(generator, join = true, outer = false, qualifier = None, attributes, logicalPlan)
```

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Where is `attributes` defined? I mean:
```
Generate(generator, join = true, outer = false, qualifier = None, generatorOutput = Nil, logicalPlan)
```

Comment (navis) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Done. Also changed generatorOutput to null in explode(string, string).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 3]
```diff
@@ -53,7 +53,7 @@ trait Generator extends Expression {
    * The output element data types in structure of Seq[(DataType, Nullable)]
    * TODO we probably need to add more information like metadata etc.
    */
-  def elementTypes: Seq[(DataType, Boolean)]
+  def elementTypes: Seq[(DataType, Boolean, Option[String])]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Seems we always set the name of elements? When will we use `None`?

Comment (navis) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Right, we can remove Option for field name. I'll update that.

**Comment by cloud-fan** [comment]
LGTM   cc @yhuai

**Comment by yhuai** [comment]
Thanks! Merging to master.

**Pull request closed without merging**